### PR TITLE
Dump rho per species

### LINKED ISSUE #1037: Fix and upgrade rho in diags
Currently, the charge deposition diagnostics contains `rho` from the previous timestep, which is not practical. Besides, when a user requests rho in the diagnostics, rho is computed at every timestep, which is not needed. A clean way to fix it would be to:
 - Avoid computing rho every timestep (delete `WarpX::plot_rho` handling in the code)
 - Re-compute `rho` on the fly from the diagnostics. Hence, in `Diagnostics.cpp`, instead of
```c++
} else if ( m_varnames[comp] == "rho" ){
    m_all_field_functors[lev][comp] = std::make_unique<CellCenterFunctor>(warpx.get_pointer_rho_fp(lev), lev, m_crse_ratio);
```
we could have
```c++
} else if ( m_varnames[comp] == "rho" ){
    m_all_field_functors[lev][comp] = std::make_unique<RhoFunctor>(lev, m_crse_ratio);
```
where functor `RhoFunctor` would
    - Allocate a node-centered `tmp` MultiFab (so that we do exactly the same is in the simulation)
    - Deposit rho in MultiFab `tmp`
    - Cell-center `tmp` to `Diagnostics::m_mf_output`.
Examples of this type of procedures can be found in `PartPerCellFunctor`, `DivEFunctor` and `DivBFunctor`. This should give the same result as the original implementation to machine precision, allowing for easy debugging.

As a second step, we could also make `RhoFunctor` compute per-species charge deposition, and store the results in different components of `Diagnostics::m_mf_output`. For an example of a single functor writing into several components of the output MultiFab `Diagnostics::m_mf_output`, see option `ncomp` in `CellCenterFunctor`, as used in `Diagnostics::AddRZModesToDiags`.

### PR DESCRIPTION
**Pull request by EZoni**
**Goal**
Add the capability to dump rho per species for selected species and close #1037.

**Description**
Some features are missing, e.g. back-transformed diagnostics (I believe we need to start using `RhoFunctor` there, first).
The current implementation selects the species to dump rho based on the input parameter `diag_name.species_name.variables`.
For instance,
```
diag.fields_to_plot = Ex Ez rho
diag.electrons.variables = w ux uz rho
diag.positrons.variables = w ux uz
```
will dump the total rho and the rho for electrons only. To avoid dumping the total rho, one can remove `rho` from the list `diag_name.fields_to_plot`.

**Tests**
This is an example from the 2D test Langmuir_multi_2d_psatd from our test suite:
<p align="center">
    
    
</p>
<p align="center">
    
    
</p>

### CODE REVIEW

**Comment by ax3l** [comment]
@EZoni sorry, one more update from `development`, please :)

**Comment by EZoni** [comment]
@ax3l Done. The Travis build errored due to the usual time-outs. Here I have two plots obtained by running this branch on the input file that you are adding in #1204, in the low resolution configuration. Do they look reasonable to you?

<p align="center">
    
    
</p>

**Comment by ax3l** [comment]
@EZoni they look great, maybe we should do a lineout along the longitudinal axis in a lin-log plot (lin. z, log. density).
(For early times, we would see the debye sheet and 2-omega imprint better in such a plot.)

In the PR description:
> This is an example from a 2D test taken from our test suite

Can you please note which inputs those were?

File: Source/Diagnostics/Diagnostics.cpp [hunk 1]
```diff
@@ -103,6 +103,57 @@ Diagnostics::BaseReadParameters ()
 
     bool species_specified = pp.queryarr("species", m_species_names);
 
+    // Prepare to dump rho per species
+    amrex::Vector<std::string> variables;
```
Comment (ax3l) [inline Source/Diagnostics/Diagnostics.cpp]: Please add the following includes above:
```C++
#include <AMReX_Vector.H>

#include <string>
```

`variables` is a quite generic name, could we name this something like `species_variables`?
```suggestion
    amrex::Vector<std::string> species_variables;  // variable selection per species
```

File: Source/Diagnostics/Diagnostics.cpp [hunk 2]
```diff
@@ -103,6 +103,57 @@ Diagnostics::BaseReadParameters ()
...
+    const MultiParticleContainer& mpc = warpx.GetPartContainer();
+    // If diag.species is not provided, loop over all species to check whether
+    // rho per species is requested
+    if (m_species_names.size() == 0) m_species_names = mpc.GetSpeciesNames();
```
Comment (ax3l) [inline Source/Diagnostics/Diagnostics.cpp]:
```suggestion
    if (m_species_names.size() == 0u) m_species_names = mpc.GetSpeciesNames();
```

File: Source/Diagnostics/Diagnostics.cpp [hunk 3]
```diff
@@ -103,6 +103,57 @@ Diagnostics::BaseReadParameters ()
...
+    // ns: total number of species
+    // ns_dump_rho: number of species that dump rho per species
+    const int ns = int(m_species_names.size());
+    int ns_dump_rho = 0;
+    // is: species index to loop over all species
```
Comment (ax3l) [inline Source/Diagnostics/Diagnostics.cpp]: let's simplify this a little:
```suggestion
```

File: Source/Diagnostics/Diagnostics.cpp [hunk 4]
```diff
@@ -103,6 +103,57 @@ Diagnostics::BaseReadParameters ()
...
+    // is_dump_rho: species index to loop over species that dump rho per species
+    int is;
+    int is_dump_rho;
+    // Loop over all species
+    for (is = 0; is < ns; is++) {
```
Comment (ax3l) [inline Source/Diagnostics/Diagnostics.cpp]:
```suggestion
    for (int is = 0; is < ns; is++) {
```

File: Source/Diagnostics/Diagnostics.cpp [hunk 5]
```diff
@@ -103,6 +103,57 @@ Diagnostics::BaseReadParameters ()
...
+            }
+        }
+        // Clear content of variables before moving to the next species
+        variables.clear();
+    }
+    // Allocate array of species indices that dump rho per species
+    m_rho_per_species_index.resize(ns_dump_rho);
+    is_dump_rho = 0;
+    // Loop over all species
+    for (is = 0; is < ns; is++) {
```
Comment (ax3l) [inline Source/Diagnostics/Diagnostics.cpp]:
```suggestion
    for (int is = 0; is < ns; is++) {
```

File: Source/Diagnostics/Diagnostics.cpp (hunk 4 above)
Comment (ax3l) [reply Source/Diagnostics/Diagnostics.cpp]: You can also make this a nice range-based loop:
```suggestion
    for (const auto & species_name : m_species_names) {
```

File: Source/Diagnostics/Diagnostics.cpp [hunk 6]
```diff
@@ -103,6 +103,57 @@ Diagnostics::BaseReadParameters ()
...
+    const int ns = int(m_species_names.size());
+    int ns_dump_rho = 0;
+    // is: species index to loop over all species
+    // is_dump_rho: species index to loop over species that dump rho per species
+    int is;
+    int is_dump_rho;
+    // Loop over all species
+    for (is = 0; is < ns; is++) {
+        // Parse input argument <diag_name>.<species_name>.variables
+        amrex::ParmParse pp_sp(m_diag_name + "." + m_species_names[is]);
```
Comment (ax3l) [inline Source/Diagnostics/Diagnostics.cpp]:
```suggestion
        amrex::ParmParse pp_sp(m_diag_name + "." + species_name);
```

File: Source/Diagnostics/Diagnostics.cpp [hunk 7]
```diff
@@ -103,6 +103,57 @@ Diagnostics::BaseReadParameters ()
...
+        if (pp_sp.queryarr("variables", variables)) {
+            for (const auto& var : variables) {
+                if (var == "rho") {
+                    // Add strings "rho_<species_name>" to m_varnames
+                    m_varnames.push_back("rho_" + m_species_names[is]);
```
Comment (ax3l) [inline Source/Diagnostics/Diagnostics.cpp]:
```suggestion
                    m_varnames.push_back("rho_" + species_name);
```

File: Source/Diagnostics/Diagnostics.cpp [hunk 8]
```diff
@@ -103,6 +103,57 @@ Diagnostics::BaseReadParameters ()
 
     bool species_specified = pp.queryarr("species", m_species_names);
 
+    // Prepare to dump rho per species
```
Comment (ax3l) [inline Source/Diagnostics/Diagnostics.cpp]: I feel this is a functional block (maybe even a good candidate for a member function). So we could potentially describe what it does from the beginning like this?
```suggestion
    // Prepare to dump rho per species:
    //   add rho_<species> to m_varnames if selected by the user
```

File: Source/Diagnostics/FullDiagnostics.cpp [hunk 9]
```diff
@@ -351,6 +351,9 @@ FullDiagnostics::InitializeFieldFunctors (int lev)
     // Clear any pre-existing vector to release stored data.
     m_all_field_functors[lev].clear();
 
+    // Species index to loop over species that dump rho per species
+    int is_dump_rho = 0;
```
Comment (ax3l) [inline Source/Diagnostics/FullDiagnostics.cpp:355]: This shared state might be a bit too implicit to quickly follow and keep track of. Can we derive the final value in `is_dump_rho` from something else, e.g. `m_rho_per_species_index.size()` or the count of non-zero entries in `m_rho_per_species_index` ([`std::count_if`](https://en.cppreference.com/w/cpp/algorithm/count))?

File: Source/Diagnostics/Diagnostics.cpp [hunk 10]
```diff
@@ -103,6 +103,57 @@ Diagnostics::BaseReadParameters ()
...
+                    // Count number of species that dump rho per species
+                    ns_dump_rho++;
+                }
+            }
+        }
+        // Clear content of variables before moving to the next species
+        variables.clear();
+    }
+    // Allocate array of species indices that dump rho per species
+    m_rho_per_species_index.resize(ns_dump_rho);
```
Comment (ax3l) [inline Source/Diagnostics/Diagnostics.cpp:142]: Is this sufficiently save or should one also assign a value to all entries (e.g. `0`)?

File: Source/Diagnostics/Diagnostics.cpp [hunk 11]
```diff
@@ -103,6 +103,57 @@ Diagnostics::BaseReadParameters ()
...
+    is_dump_rho = 0;
+    // Loop over all species
+    for (is = 0; is < ns; is++) {
+        // Parse input argument <diag_name>.<species_name>.variables
+        amrex::ParmParse pp_sp(m_diag_name + "." + m_species_names[is]);
+        if (pp_sp.queryarr("variables", variables)) {
+            for (const auto& var : variables) {
+                if (var == "rho") {
+                    // Fill array of species indices that dump rho per species
+                    m_rho_per_species_index[is_dump_rho] = is;
```
Comment (ax3l) [inline Source/Diagnostics/Diagnostics.cpp]: Whenever it is not performance critical, the range-checked `.at()` is what I personally prefer for long-term stability:
```suggestion
                    m_rho_per_species_index.at(is_dump_rho) = is;
```

File: Source/Diagnostics/ParticleDiag/ParticleDiag.cpp [hunk 12]
```diff
@@ -32,13 +32,17 @@ ParticleDiag::ParticleDiag(std::string diag_name, std::string name, WarpXParticl
...
         for (const auto& var : variables){
-            // Return error if var not in PIdx.
-            WarpXUtilMsg::AlwaysAssert(
-                ParticleStringNames::to_index.count(var),
-                "ERROR: variables argument '" + var +
-                "' not in ParticleStringNames"
-                );
-            plot_flags[ParticleStringNames::to_index.at(var)] = 1;
+            // The string "rho" is needed to dump rho per species, but does not
+            // need to be added to the PIdx structure, hence we skip the check
```
Comment (ax3l) [inline Source/Diagnostics/ParticleDiag/ParticleDiag.cpp]: Idea:
```suggestion
            // The string "rho" is needed to dump rho per species, which is
            // generated on the fly from existing species variables (position, weight).
            // Hence, "rho" is not part of the species' PIdx variables.
```

File: Source/Diagnostics/FullDiagnostics.cpp (hunk 9 above)
Comment (ax3l) [reply Source/Diagnostics/FullDiagnostics.cpp:355]: Clarified: I overlooked that this a temporary variable, no member variable.

File: Source/Diagnostics/Diagnostics.cpp (hunk 10 above)
Comment (ax3l) [reply Source/Diagnostics/Diagnostics.cpp:142]: Probably sufficiently safe since `::BaseReadParameters` is only read once.

**Comment by EZoni** [comment]
<p align="center">
    
    
</p>

**Comment by EZoni** [comment]
@ax3l Sorry for closing the PR by mistake. Anyways, same plots for the laser ion acceleration case obtained with the openPMD format:
<p align="center">
        
        
</p>

**Comment by ax3l** [comment]
Awesome, this is wonderful :star_struck:

**Comment by RevathiJambunathan** [comment]
@EZoni amazing PR! Thank you for this!

**Comment by EZoni** [comment]
@RevathiJambunathan Thank you! Actually, I have one question for you: do you know what needs to be done in order to use `RhoFunctor` instead of `CellCenterFunctor` with back-transformed diagnostics? https://github.com/ECP-WarpX/WarpX/blob/08b7bcc549ccdc412cfb06b278a92b4e0b13ead9/Source/Diagnostics/FullDiagnostics.cpp#L376-L387
It is not directly related to this PR, but I think it is something that we should have in the WarpX to-do list and I wonder if we could maybe summarize in an issue what steps need to be taken in order to complete the task.

**Comment by RevathiJambunathan** [comment]
@EZoni yes good point. I started thinking of BTD as well, when I saw this PR.

In BTDiagnostics.cpp, we first generated cell-centered MultiFab with Ex,Ey,Ez,Bx,By,Bz,jx,jy,jz,and rho.
This is where we use the RhoFunctor. https://github.com/ECP-WarpX/WarpX/blob/08b7bcc549ccdc412cfb06b278a92b4e0b13ead9/Source/Diagnostics/BTDiagnostics.cpp#L340

```
    // cell-center functors for BackTransform Diags
    for (int comp=0, n=m_cell_center_functors[lev].size(); comp<n; comp++){
        if        ( m_cellcenter_varnames[comp] == "Ex" ){
            m_cell_center_functors[lev][comp] = std::make_unique<CellCenterFunctor>(warpx.get_pointer_Efield_aux(lev, 0), lev, m_crse_ratio);
        } else if ( m_cellcenter_varnames[comp] == "Ey" ){
            m_cell_center_functors[lev][comp] = std::make_unique<CellCenterFunctor>(warpx.get_pointer_Efield_aux(lev, 1), lev, m_crse_ratio);
        } else if ( m_cellcenter_varnames[comp] == "Ez" ){
            m_cell_center_functors[lev][comp] = std::make_unique<CellCenterFunctor>(warpx.get_pointer_Efield_aux(lev, 2), lev, m_crse_ratio);
        } else if ( m_cellcenter_varnames[comp] == "Bx" ){
            m_cell_center_functors[lev][comp] = std::make_unique<CellCenterFunctor>(warpx.get_pointer_Bfield_aux(lev, 0), lev, m_crse_ratio);
        } else if ( m_cellcenter_varnames[comp] == "By" ){
            m_cell_center_functors[lev][comp] = std::make_unique<CellCenterFunctor>(warpx.get_pointer_Bfield_aux(lev, 1), lev, m_crse_ratio);
        } else if ( m_cellcenter_varnames[comp] == "Bz" ){
            m_cell_center_functors[lev][comp] = std::make_unique<CellCenterFunctor>(warpx.get_pointer_Bfield_aux(lev, 2), lev, m_crse_ratio);
        } else if ( m_cellcenter_varnames[comp] == "jx" ){
            m_cell_center_functors[lev][comp] = std::make_unique<CellCenterFunctor>(warpx.get_pointer_current_fp(lev, 0), lev, m_crse_ratio);
        } else if ( m_cellcenter_varnames[comp] == "jy" ){
            m_cell_center_functors[lev][comp] = std::make_unique<CellCenterFunctor>(warpx.get_pointer_current_fp(lev, 1), lev, m_crse_ratio);
        } else if ( m_cellcenter_varnames[comp] == "jz" ){
            m_cell_center_functors[lev][comp] = std::make_unique<CellCenterFunctor>(warpx.get_pointer_current_fp(lev, 2), lev, m_crse_ratio);
        } else if ( m_cellcenter_varnames[comp] == "rho" ){
            m_cell_center_functors[lev][comp] = std::make_unique<RhoFunctor>(lev, m_crse_ratio);
        }
    }
```

Note that  -- regarless of the type of user-defined field for BTD, the cell-centered MultiFab will always have these ten components.

**Review by ax3l — APPROVED** [review]
:rocket: :sparkles:

File: Docs/source/running_cpp/parameters.rst [hunk 13]
```diff
@@ -1432,13 +1432,16 @@ s disabled.
...
 
 * ``<diag_name>.<species_name>.variables`` (list of `strings` separated by spaces, optional)
-    List of particle quantities to write to output file.
-    By defaults, all quantities are written to file. Choices are
+    List of particle quantities or species-specific field quantities to write to output file.
+    Choices are
 
     * ``w`` for the particle weight,
 
     * ``ux`` ``uy`` ``uz`` for the particle momentum,
```
Comment (ax3l) [inline Docs/source/running_cpp/parameters.rst:1440]: @EZoni I just realized: isn't this list missing: `x`, `y`, `z` or does this only refer to the SoA variables?

Comment (EZoni) [reply Docs/source/running_cpp/parameters.rst:1440]: I'm not sure about that. It seems to refer to this namespace: 
https://github.com/ECP-WarpX/WarpX/blob/b1c9e243f1a00f597dfaaa122597c015d71b67ac/Source/Particles/WarpXParticleContainer.H#L62-L73
I think that the input argument that the doc refers to, `<diag_name>.<species_name>.variables`, is read here:
https://github.com/ECP-WarpX/WarpX/blob/b1c9e243f1a00f597dfaaa122597c015d71b67ac/Source/Diagnostics/ParticleDiag/ParticleDiag.cpp#L11-L14
I also thought we could store the particle positions, but this `variables` input seems to be matched with the elements in the namespace `ParticleStringNames` above (which does not include `x`, `y` and `z`):
https://github.com/ECP-WarpX/WarpX/blob/b1c9e243f1a00f597dfaaa122597c015d71b67ac/Source/Diagnostics/ParticleDiag/ParticleDiag.cpp#L33-L48

Comment (EZoni) [reply Docs/source/running_cpp/parameters.rst:1440]: There is an additional comment in the same part of the doc that you pointed at, stating that the particle positions are always included. I guess that is why they were not mentioned among the optional arguments.

Comment (ax3l) [reply Docs/source/running_cpp/parameters.rst:1440]: Yes, I think that's the reason. Particle positions (and id) are part of the AoS data, the variables here describe additional SoA data.

**Pull request merged**